### PR TITLE
Pytest docking testing and ability to use a single receptor `oedu`

### LINKED ISSUE #222: pytest based testing for docking workflow.
> @hmacdope I kinda think we should leave further improvements of the unit tests to a future. I don't really have the bandwidth to make new changes on that right now and would like to get these improved logging options added

No worries, probably a good call. I will siphon it off to a different PR.

_Originally posted by @hmacdope in https://github.com/choderalab/asapdiscovery/issues/192#issuecomment-1504281419_

### PR DESCRIPTION
**Pull request by hmacdope**
Fixes #222 

- Adds pytest harness for docking to know when we have broken it (eg it is currently broken on `main @ 095b4e1e`)
- Adds ability to specify a single `.oedu` file rather than a glob or list (was previously insisting a single file be a CSV). 
- Adds a few error checks.

### CODE REVIEW

**Review by kaminow — COMMENTED** [review]
just a couple questions on some changes to the docking script

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 1]
```diff
@@ -246,22 +248,33 @@ def parse_du_filenames(receptors, regex, basefile="predocked.oedu"):
...
+                all_fns = [
+                    os.path.join(os.path.dirname(fn), basefile)
+                    for fn in df["Docked_File"]
+                ]
+            except KeyError:
+                raise ValueError("Docked_File column not found in given CSV file.")
+        elif file_extn == ".oedu":
+            all_fns = [receptors]
+            if by_compound:
+                raise ValueError("Cannot use with a single DesignUnit file.")
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:268]: The help to `-r` seemed to suggest that if by_compound is true then we expect a  CSV only. Let me know if this isn't true anymore and I can remove the error. 

```
    parser.add_argument(
        "-r",
        "--receptor",
        required=True,
        nargs="+",
        help=(
            "Path/glob to prepped receptor(s), or best_results.csv file if "
            "--by_compound is given."
        ),
    )

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 2]
```diff
@@ -246,22 +248,33 @@ def parse_du_filenames(receptors, regex, basefile="predocked.oedu"):
...
+            df = pandas.read_csv(receptors)
+            try:
+                all_fns = [
+                    os.path.join(os.path.dirname(fn), basefile)
+                    for fn in df["Docked_File"]
+                ]
+            except KeyError:
+                raise ValueError("Docked_File column not found in given CSV file.")
+        elif file_extn == ".oedu":
+            all_fns = [receptors]
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:266]: @kaminow  agree in principle but the `glob` is  on the other branch of the ` if os.isfile(receptors)` branch which is why we were having some issues,. 

Ie if you provide a path to a full file, then it was always trying to read it as a CSV and the glob was not being hit. I know its a bit verbose but I think allows for more debugging/logging.

[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:267]: ah yep good point, my bad

**Review by apayne97 — APPROVED** [review]
LGTM! Looking forward to using this testing framework to implement other tests!

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaminow) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:280]: sorry for the delayed response, but I think what happened here is I was using the `--by_compound` arg to indicate a different directory structure for the `.oedu` files, but at this point I don't think it actually matters. I think we should keep the `--by_compound` CLI arg to indicate which regex to use, but we can probably change the above help text to something like "Path/glob to prepped receptor(s), or CSV file containing receptor paths.", and get rid of the `ValueError` in this code. happy to implement these changes if you want, just don't want to mess with your branch without warning

File: asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py [hunk 3]
```diff
@@ -255,15 +259,27 @@ def parse_du_filenames(receptors, regex, log_name, basefile="predocked.oedu"):
             for fn in files
             if fn[-4:] == "oedu"
         ]
-    elif os.path.isfile(receptors):
```
[the remark this answers is not in the document]
Comment (hmacdope) [reply asapdiscovery-docking/asapdiscovery/docking/scripts/run_docking_oe.py:258]: good spotting.

**Comment by hmacdope** [comment]
@kaminow  hopefully addressed comments :)

**Review by kaminow — APPROVED** [review]
looks great now, thanks!

**Pull request merged**